tests/test_base.py

```diff
@@ -312,3 +312,24 @@ class Node(Base):
 
     model = Node(id="a", metadata={"b": np.array([0])})
     model.to_json()
+
+
+def test_ndarray_bson_metadata():
+    import numpy as np
+
+    @modelspec.define(eq=False)
+    class Node(Base):
+        id: str = field(validator=instance_of(str))
+        metadata: Optional[Dict[str, Any]] = field(
+            kw_only=True, default=None, validator=optional(instance_of(dict))
+        )
+
+    model = Node(
+        id="a", metadata={"b": np.random.randint(10, 10000, 1000)}
+    )  # creates an array of 1000 integers ranging from 1 to 10000
+    model.to_bson_file("sample.bson")
+    model.to_json_file("sample.json")  # to compare the size
+    model.from_bson_file("sample.bson")
+
+
+test_ndarray_bson_metadata()
```

Comment:
Also try to add a more complete test. Something that handles multiple types of ndarray and list combinations. Maybe:

```python
def test_bson_array(tmp_path):
    test_filename = str(tmp_path / "test_array.bson")

    @modelspec.define
    class ArrayTestModel(Base):
        array: np.ndarray = field(default=None)
        list_of_lists: List[List[int]] = field(default=None)
        ragged_list: List[List[int]] =field(default=None)

    m = ArrayTestModel(array=np.arange(27).reshape((3,3,3)), list_of_lists=[[1.2], [3.4]], ragged_list=[[1,2], [1]])

    m.to_bson_file(test_filename)

    # Load it back in
    m2 = ArrayTestModel.from_bson_file(test_filename)

    # Check we get the same values back
    assert m.array == m2.array
    assert m.list_of_lists == m2.list_of_lists
    assert m.ragged_list == m2.ragged_list
```

The checks at the bottom might need to be changed to work correctly but hopefully you get what I mean. I think the current behavior is that numpy arrays are serialized as lists of lists. Also note the generation of files to a temp directory using the tmp_path fixture from pytest. This is good practice to avoid polluting the repo with generated files from tests.

Reply:
Hi @davidt0x,
Firstly the test cases provided for **array** field results in an error (tested for both json and bson):
- The to_dict fails as `cattr.Converter().unstructure is not able to **unstructure()**`
So I guess the test case is too complex for the existing architecture

Secondly (rafter removing the array field)
```
list_of_lists: List[List[int]] = field(default=None)
ragged_list: List[List[int]] =field(default=None)
```
 the data type provided while declaring "int" must match with the data type of the values provided while creating the object "float" so that the assertion results in positive.

Reply:
Hey @parikshit14 ,

My bad, there actually seems to be a bug in cattrs for handling np.ndarray using its code generation stuff on attributes that have a default value. I will file a bug with cattrs or see if I can fix and put a PR into the project.

We can work around by dropping the default for the array attribute. Also, the floats in the lists of ints was a typo, should have been commas :) Here try this:

```
@modelspec.define(eq=False)
class ArrayTestModel(Base):
    array: Optional[ValueExprType] = field()
    list_of_lists: List[List[int]] = field(default=None)
    ragged_list: List[List[int]] = field(default=None)

m = ArrayTestModel(array=np.arange(27).reshape((3,3,3)), list_of_lists=[[1, 2], [3, 4]], ragged_list=[[1,2], [1]])

m.to_json_file(test_filename)

# Load it back in
m2 = ArrayTestModel.from_json_file(test_filename)

# Check we get the same values back
assert m.array.tolist() == m2.array
assert m.list_of_lists == m2.list_of_lists
assert m.ragged_list == m2.ragged_list
```